### PR TITLE
SONARJAVA-3345 S3518 (division by zero) should handle BigDecimal and BigInteger

### PR DESCRIPTION
**Pull request by quentin-jaquier-sonarsource**

### CODE REVIEW

File: java-frontend/src/test/files/se/DivisionByZeroCheck.java [hunk 1]
```diff
@@ -495,3 +498,67 @@ private int g(int x, int y) {
...
+    BigInteger v1 = BigInteger.valueOf(42);
+    BigInteger z2 = z1.multiply(v1);
+    r = r.remainder(z2); // Noncompliant
+  }
+
+  void simpleBigInt(BigInteger r, int value) {
+    BigInteger z1 = BigInteger.valueOf(value);
+    r = r.divide(z1); // Compliant
+  }
+}
```
Comment (alban-auzeill) [inline java-frontend/src/test/files/se/DivisionByZeroCheck.java:564]: I addition to the above use cases, we should probably also support:

- The BigInteger.ZERO and BigDecimal.ZERO identifiers
```java
  void zeroConstantBigInt(BigInteger r) {
    BigInteger z1 = BigInteger.ZERO;
    r = r.divide(z1); // Noncompliant
  }
```
- Keeping "ZERO" constraint through "valueOf" from long, double to BigInteger, BigDecimal
```java
  void valueOfBigInt(BigInteger r) {
    long z1 = 0;
    BigInteger z2 = BigInteger.valueOf(z1);
    r = r.divide(z2); // Noncompliant
  }
```
- Keeping "ZERO" constraint through BigDecimal constructor from BigInteger, double, int, long
```java
  void constructorBigDecimal(BigDecimal r) {
    double z1 = 0;
    BigDecimal z2 = new BigDecimal(z1);
    r = r.divide(z2); // Noncompliant
  }
```
 - Keeping "ZERO" constraint through "toBigInteger", "toBigIntegerExact", "shortValueExact", "round", "	remainder", "pow", "longValueExact", "longValue" , "	intValueExact", "intValue", "floatValue", "doubleValue", "byteValueExact", "abs", 	shiftLeft, shiftRight
 - Support 	BigDecimal.divideToIntegralValue like BigDecimal.divide

Comment (quentin-jaquier-sonarsource) [reply java-frontend/src/test/files/se/DivisionByZeroCheck.java:738]: Good point! I added one commit for each case that you stated.

Comment (alban-auzeill) [reply java-frontend/src/test/files/se/DivisionByZeroCheck.java:749]: Nice job!
Unfortunately, now we raise a FP on the following code:
```java
    BigDecimal x = new BigDecimal(0);
    BigDecimal y = new BigDecimal(0);
    if (x == y) { // false-positive "always true"
      System.out.println("1");
    } else {
      System.out.println("2"); // <==
    }
```

**Comment by quentin-jaquier-sonarsource** [comment]
As discussed, we decided to remove support for `BigDecimal` constructor, as it introduces unexpected behavior in the symbolic execution without a lot of added value since we have rules discouraging to use such constructor: [RSPEC-2111](https://rules.sonarsource.com/java/RSPEC-2111) and [RSPEC-2129](https://rules.sonarsource.com/java/RSPEC-2129)

**Pull request merged**